hack/scripts/provision.sh

```diff
@@ -0,0 +1,996 @@
+}
+
+## DIRECT_LVM
+#
+#
+do_all_direct_lvm() {
+	local disk="$1"
+	local thinpool="$2"
+
+	say "Setting up direct_lvm thinpool $thinpool"
+
+	if [[ -z "$disk" ]]; then
+		say_warn "WARNING: -d/--disk has not been set. If you continue, the" \
+			"script will attempt to detect a free disk for formatting. Any data" \
+			"will be lost."
+		get_user_confirmation "Are you sure you wish to continue? (y/n) " || die "Aborted."
+
+		disk=$(find_free_disk || die "Could not detect free disk")
+	fi
+
+	local disk_path="/dev/$disk"
+
+	say "Will use $disk_path for direct-lvm thinpool $thinpool"
+	say_warn "All existing data on $disk_path will be overwritten."
+	get_user_confirmation || die "Aborted."
+
+	create_physical_volume "$disk_path"
+	create_volume_group "$disk_path" "$thinpool"
+	create_logical_volume "$thinpool"
+	apply_lvm_profile "$thinpool"
+	monitor_lvm_profile "$thinpool" || die "failed to monitor lvm profile"
+
+	say "Thinpool $thinpool-thinpool is ready for use"
+}
+
+# Naively find a space block device which is not in use
+# This is really unsafe as it only looks at lsblk to decide anything
+# It is a much better idea to use the --disk flag and pass in something you
+# know will be available
+find_free_disk() {
+	disks=("$(lsblk -o NAME,TYPE,MOUNTPOINT | awk '/disk/ {print $1}')")
+	root_parts=("$(lsblk -o NAME,TYPE,MOUNTPOINT | awk '/part \// { gsub("[^a-zA-Z0-9]+", "", $1); print $1}')")
+
+	# shellcheck disable=SC2068
+	for d in ${disks[@]}; do
+		for p in ${root_parts[@]}; do
+			if [[ $p != *"$d"* ]]; then
+				echo "$d" && return 0
+			fi
+		done
+	done
+
+	return 1
+}
+
+# Create a physical volume on the given device
+create_physical_volume() {
+	local disk_path="$1"
+
+	# if already exists, do nothing
+	if [[ $(pvdisplay 2>/dev/null) != *"$disk_path"* ]]; then
+		pvcreate -q "$disk_path" || die "failed to create physical volume on $disk_path"
+		say "Created physical volume on $disk_path"
+		return 0
+	fi
+
+	say "Physical volume on $disk_path already exists"
+}
+
+# Create a volume group on the given device for the thinpool
+create_volume_group() {
+	local disk_path="$1"
+	local thinpool="$2"
+
+	# if already exists, do nothing
+	if [[ $(vgdisplay 2>/dev/null) != *"$thinpool"* ]]; then
+		vgcreate -q "$thinpool" "$disk_path" || die "failed to create volume group on $disk_path"
+		say "Created volume group on $disk_path"
+		return 0
+	fi
+
+	say "Volume group on $disk_path already exists"
+}
+
+# Format the volume for thinpool storage
+create_logical_volume() {
+	local volume_group="$1"
+
+	# if already exists, do nothing
+	if [[ $(lvdisplay 2>/dev/null) != *"$volume_group"* ]]; then
+		lvcreate -q --wipesignatures y -n thinpool "$volume_group" -l 95%VG || die "Failed to create logical volume for thinpool data"
+		say "Created logical volume for $volume_group thinpool data"
+
+		lvcreate -q --wipesignatures y -n thinpoolmeta "$thinpool" -l 1%VG || die "Failed to create logical volume for thinpool metadata"
```

Comment:
Hmm.. That `95%` and `1%` combo seems strange.
1. where is the remaining `4%`
2. the `METADATA_SPARSE_SIZE="10G"` `DATA_SPARSE_SIZE="100G"` pair has a 1:10 ratio.

Reply:
It's even more fishy, we have fix size for spare files, but we use dynamic volume size with lvm.

If the disk full size is less than `$((DATA_SPARSE_SIZE + METADATA_SPARSE_SIZE))` we should fail before we even try to erase it.

Reply:
nvm, i see `DATA_SPARSE_SIZE` and `METADATA_SPARSE_SIZE` are used for dev only without lvm.

(leave this thread here to record my confusion :laughing: )

Reply:
i copied all the devpool stuff from our existing script, did not actually read any of it 😁